### PR TITLE
feat(statedb): support loading from checkpoint state

### PR DESCRIPTION
**Pull request by zeroqn**
Add ```CheckPoint```, ```SubState``` and ```StateDBMode```
Save every withdrawal and transaction post state to db
Able to load statedb from checkpoint(block_number with ```SubState```) and ```StateDBMode```

### CODE REVIEW

File: crates/block-producer/src/produce_block.rs [hunk 1]
```diff
@@ -76,9 +76,10 @@ pub fn produce_block(param: ProduceBlockParam<'_>) -> Result<ProduceBlockResult>
     let state_db = {
         let tip_block_hash = db.get_tip_block_hash()?;
         assert_eq!(parent_block_hash, tip_block_hash);
-        StateDBTransaction::from_version(
+        StateDBTransaction::from_checkpoint(
             &db,
-            StateDBVersion::from_history_state(&db, tip_block_hash, None)?,
+            CheckPoint::from_block_hash(&db, tip_block_hash, SubState::Block)?,
+            StateDBMode::Write,
```
Comment (jjyr) [inline crates/block-producer/src/produce_block.rs:82]: StateDBMode::ReadOnly,

**Review by jjyr — CHANGES_REQUESTED** [review]
Please add a test case of state-DB with a withdrawal sub-state.

File: crates/store/src/transaction.rs [hunk 2]
```diff
@@ -267,6 +279,7 @@ impl StoreTransaction {
         committed_info: packed::L2BlockCommittedInfo,
         global_state: packed::GlobalState,
         tx_receipts: Vec<packed::TxReceipt>,
+        post_states: Vec<packed::AccountMerkleState>,
```
Comment (jjyr) [inline crates/store/src/transaction.rs:282]:
```suggestion
        withdrawal_receipts: Vec<packed::WithdrawalReceipt>,
```

**Review by jjyr — CHANGES_REQUESTED** [review]

**Comment by zeroqn** [comment]
All Done

File: crates/store/src/state_db.rs [hunk 3]
```diff
@@ -163,94 +272,93 @@ impl<'db> StateDBTransaction<'db> {
...
         Ok(account_merkle_state)
     }
 
-    fn get_key_with_suffix(&self, key: &[u8]) -> Vec<u8> {
-        let (block_number, tx_index) = self.version.extract_block_number_and_index_number();
-        [key, &block_number.to_be_bytes(), &tx_index.to_be_bytes()].concat()
+    fn get_key_with_suffix(&self, key: &[u8]) -> Result<Vec<u8>, Error> {
+        let (block_number, index) = self
+            .checkpoint
+            .extract_block_number_and_index_number(&self.inner, self.mode)?;
```
Comment (jjyr) [inline crates/store/src/state_db.rs:353]: We can use `.expect(...)` to eliminate the error here.

File: crates/store/src/state_db.rs [hunk 4]
```diff
@@ -437,3 +563,109 @@ impl<'a, 'db> CodeStore for StateTree<'a, 'db> {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (jjyr) [inline crates/store/src/state_db.rs:568]: Put tests into crates/store/src/tests/state_db.rs

File: crates/store/src/transaction.rs [hunk 5]
```diff
@@ -11,14 +11,28 @@ use gw_db::schema::{
...
-    packed::{self, RollupConfig, TransactionKey},
+    packed::{self, Byte32, RollupConfig, TransactionKey},
     prelude::*,
 };
 use std::{borrow::BorrowMut, collections::HashMap};
 
 const NUMBER_OF_CONFIRMATION: u64 = 100;
 
+#[derive(Clone)]
+pub struct WithdrawalReceipt {
```
Comment (jjyr) [inline crates/store/src/transaction.rs:24]: Put this into gw-types/store.mol

**Review by jjyr — CHANGES_REQUESTED** [review]

**Review by jjyr — APPROVED** [review]

**Pull request merged**